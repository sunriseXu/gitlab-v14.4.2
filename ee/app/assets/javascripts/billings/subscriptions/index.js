import Vue from 'vue';
import Vuex from 'vuex';
import SubscriptionApp from './components/app.vue';
import initialStore from './store';

Vue.use(Vuex);

export default (containerId = 'js-billing-plans') => {
  const containerEl = document.getElementById(containerId);

  if (!containerEl) {
    return false;
  }

  const {
    namespaceId,
    namespaceName,
    addSeatsHref,
    planRenewHref,
    customerPortalUrl,
    billableSeatsHref,
    planName,
    refreshSeatsHref,
    action,
    trialPlanName,
  } = containerEl.dataset;

  return new Vue({
    el: containerEl,
    store: new Vuex.Store(initialStore()),
    provide: {
      namespaceId: Number(namespaceId),
      namespaceName,
      addSeatsHref,
      planRenewHref,
      customerPortalUrl,
      billableSeatsHref,
      planName,
      refreshSeatsHref,
      availableTrialAction: action,
      trialPlanName,
    },
    render(createElement) {
      return createElement(SubscriptionApp);
    },
  });
};
