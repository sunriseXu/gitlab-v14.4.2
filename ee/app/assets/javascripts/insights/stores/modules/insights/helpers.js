import { CHART_TYPES } from 'ee/insights/constants';
import { __, s__ } from '~/locale';

const getGroupByValue = (queryParams) => {
  return queryParams.group_by || queryParams.params.group_by;
};

const getTypeValue = (queryParams) => {
  return queryParams.issuable_type || queryParams.params.issuable_type || queryParams.params.metric;
};

const getAxisTitle = (label) => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    case 'deployment_frequency':
      return s__('DORA4Metrics|Deployment frequency');
    case 'lead_time_for_changes':
      return s__('DORA4Metrics|Lead time for changes (median days)');
    case 'time_to_restore_service':
      return s__('DORA4Metrics|Time to restore service (median days)');
    case 'change_failure_rate':
      return s__('DORA4Metrics|Change failure rate (percentage)');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: queryParams },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(getGroupByValue(queryParams)),
    yAxisTitle: getAxisTitle(getTypeValue(queryParams)),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
