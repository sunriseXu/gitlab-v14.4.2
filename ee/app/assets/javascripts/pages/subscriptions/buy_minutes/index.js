import mountSubscriptionsApplication from 'ee/subscriptions/buy_minutes';
import { trackAddToCartUsageTab } from '~/google_tag_manager';

mountSubscriptionsApplication(document.getElementById('js-buy-minutes'));
trackAddToCartUsageTab();
