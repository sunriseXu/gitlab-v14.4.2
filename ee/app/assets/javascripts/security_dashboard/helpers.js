import isPlainObject from 'lodash/isPlainObject';
import groupBy from 'lodash/groupBy';
import {
  REPORT_TYPES_WITH_MANUALLY_ADDED,
  REPORT_TYPES_WITH_CLUSTER_IMAGE,
  REPORT_TYPES_ALL,
  SEVERITY_LEVELS,
} from 'ee/security_dashboard/store/constants';
import convertReportType from 'ee/vue_shared/security_reports/store/utils/convert_report_type';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';
import { SCANNER_NAMES_MAP } from '~/security_configuration/components/constants';
import { DEFAULT_SCANNER } from './constants';

const parseOptions = (obj) =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

const toolName = s__('SecurityReports|Tool');
const allToolsName = s__('ciReport|All tools');
const stateOptions = parseOptions(VULNERABILITY_STATES);
const defaultStateOptions = stateOptions.filter((x) => ['DETECTED', 'CONFIRMED'].includes(x.id));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  options: stateOptions,
  allOption: { id: 'all', name: s__('VulnerabilityStatusTypes|All statuses') },
  defaultOptions: defaultStateOptions,
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  options: parseOptions(SEVERITY_LEVELS),
  allOption: { name: s__('ciReport|All severities') },
  defaultOptions: [],
};

export const createScannerOption = (vendor, reportType) => {
  const type = reportType.toUpperCase();

  return {
    id: `${vendor}.${type}`,
    reportType: reportType.toUpperCase(),
    name: convertReportType(reportType),
    scannerIds: [],
  };
};

// This is used on the pipeline security tab, group-level report, and instance-level report. It's
// used by the scanner filter that shows a flat list of scan types (DAST, SAST, etc) with no vendor
// grouping.
export const simpleScannerFilter = {
  name: toolName,
  id: 'reportType',
  options: parseOptions(REPORT_TYPES_WITH_MANUALLY_ADDED),
  allOption: { name: allToolsName },
  defaultOptions: [],
};

export const simpleScannerFilterPipeline = {
  name: toolName,
  id: 'reportType',
  options: parseOptions(REPORT_TYPES_WITH_CLUSTER_IMAGE),
  allOption: { name: allToolsName },
  defaultOptions: [],
};

// This is used on the project-level report. It's used by the scanner filter that shows a list of
// scan types (DAST, SAST, etc) that's grouped by vendor.
export const vendorScannerFilter = {
  name: toolName,
  id: 'scanner',
  options: Object.keys(REPORT_TYPES_WITH_MANUALLY_ADDED).map((x) =>
    createScannerOption(DEFAULT_SCANNER, x),
  ),
  allOption: { name: allToolsName },
  defaultOptions: [],
};

export const activityOptions = {
  NO_ACTIVITY: { id: 'NO_ACTIVITY', name: s__('SecurityReports|No activity') },
  WITH_ISSUES: { id: 'WITH_ISSUES', name: s__('SecurityReports|With issues') },
  NO_LONGER_DETECTED: { id: 'NO_LONGER_DETECTED', name: s__('SecurityReports|No longer detected') },
};

export const activityFilter = {
  name: s__('Reports|Activity'),
  id: 'activity',
  options: Object.values(activityOptions),
  allOption: { name: s__('SecurityReports|All activity') },
  defaultOptions: [],
};

export const projectFilter = {
  name: s__('SecurityReports|Project'),
  id: 'projectId',
  options: [],
  allOption: { name: s__('ciReport|All projects') },
  defaultOptions: [],
};

export const clusterFilter = {
  name: s__('SecurityReports|Cluster'),
  id: 'cluster',
  options: [],
  allOption: { name: s__('ciReport|All clusters') },
  defaultOptions: [],
};

export const imageFilter = {
  name: s__('SecurityReports|Image'),
  id: 'image',
  options: [],
  allOption: { name: s__('ciReport|All images') },
  defaultOptions: [],
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES_ALL[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = (pageInfo) => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

/**
 * Provided a vulnerability scanners from the GraphQL API, this returns an array that is
 * formatted so it can be displayed in the dropdown UI.
 *
 * @param {Array} vulnerabilityScanners
 * @returns {Array} formatted vulnerabilityScanners
 */
export const getFormattedScanners = (vulnerabilityScanners) => {
  const groupedByReportType = groupBy(vulnerabilityScanners, 'reportType');

  return Object.entries(groupedByReportType).map(([reportType, scanners]) => {
    return {
      id: reportType,
      reportType,
      name: SCANNER_NAMES_MAP[reportType] || SCANNER_NAMES_MAP.GENERIC,
      scannerIds: scanners.map(({ id }) => id),
    };
  });
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

export default () => ({});
