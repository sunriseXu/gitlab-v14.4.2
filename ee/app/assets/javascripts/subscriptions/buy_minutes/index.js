import Vue from 'vue';
import ensureData from '~/ensure_data';
import { planTags, I18N_CI_1000_MINUTES_PLAN } from 'ee/subscriptions/buy_addons_shared/constants';
import App from 'ee/subscriptions/buy_addons_shared/components/app.vue';
import { writeInitialDataToApolloCache } from '../buy_addons_shared/utils';
import apolloProvider from '../buy_addons_shared/graphql';

export default (el) => {
  if (!el) {
    return null;
  }

  const extendedApp = ensureData(App, {
    parseData: writeInitialDataToApolloCache.bind(null, apolloProvider),
    data: el.dataset,
    shouldLog: true,
    provide: {
      tags: [planTags.CI_1000_MINUTES_PLAN],
      i18n: I18N_CI_1000_MINUTES_PLAN,
    },
  });

  return new Vue({
    el,
    name: 'BuyCIMinutes',
    apolloProvider,
    render(createElement) {
      return createElement(extendedApp);
    },
  });
};
