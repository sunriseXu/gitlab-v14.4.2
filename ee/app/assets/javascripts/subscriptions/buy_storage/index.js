import Vue from 'vue';
import ensureData from '~/ensure_data';
import { planTags, I18N_STORAGE_PLAN } from 'ee/subscriptions/buy_addons_shared/constants';
import App from 'ee/subscriptions/buy_addons_shared/components/app.vue';
import apolloProvider from '../buy_addons_shared/graphql';
import { writeInitialDataToApolloCache } from '../buy_addons_shared/utils';

export default (el) => {
  if (!el) {
    return null;
  }

  const extendedApp = ensureData(App, {
    parseData: writeInitialDataToApolloCache.bind(null, apolloProvider),
    data: el.dataset,
    shouldLog: true,
    provide: {
      tags: [planTags.STORAGE_PLAN],
      i18n: I18N_STORAGE_PLAN,
    },
  });

  return new Vue({
    el,
    name: 'BuyStorage',
    apolloProvider,
    render(createElement) {
      return createElement(extendedApp);
    },
  });
};
