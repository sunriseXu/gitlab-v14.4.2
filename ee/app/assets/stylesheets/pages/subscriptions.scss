#zuora_payment {
  margin-right: -8px;

  iframe {
    // stylelint-disable-next-line
    background-color: unset;
    margin: -4px;
  }
}

// We can't get rid of these styles yet because there is a section of the registration with subscription flow which is
// still making use of them. See https://gitlab.com/gitlab-org/gitlab/-/issues/359138 for more details.
/* Checkout Page */
$subscriptions-form-group-width: 420px;
$subscriptions-full-width-md: $subscriptions-form-group-width + $gl-padding * 2 + 2;
$subscriptions-full-width-lg: 541px;

.subscriptions-layout-html {
  .container {
    max-width: none;
    padding-top: 40px;
  }

  .checkout-pane {
    border-bottom: 1px solid $gray-100;
    flex-grow: 1;
    flex-shrink: 0;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      border-bottom: 0;
      border-right: 1px solid $border-color;
    }
  }

  .checkout-alert {
    width: 100%;
    max-width: $subscriptions-full-width-md;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      max-width: $subscriptions-full-width-lg;
    }
  }

  .checkout {
    align-items: center;
    flex-grow: 1;
    max-width: $subscriptions-full-width-md;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      justify-content: inherit !important;
      max-width: none;
    }

    .full-width {
      max-width: $subscriptions-full-width-lg;
      width: 100%;
    }
  }

  .bar {
    @media(min-width: map-get($grid-breakpoints, lg)) {
      width: 400px;
    }

    list-style-position: inside;
    list-style-type: decimal;

    .phase {
      border-bottom: 5px solid $gray-100;
      color: $gray-500;
      flex-basis: 100%;
      font-size: $gl-font-size-12;
      line-height: $gl-line-height;
      padding-bottom: $gl-padding-8;

      &.current,
      &.finished {
        border-bottom-color: $blue-400;
      }

      &.current {
        color: $gray-900;
      }
    }
  }

  .checkmark {
    color: $green-500;
    margin-top: -1px;
  }

  .card {
    margin-bottom: 0;
    min-height: 32px;
    padding: $gl-padding;
    max-width: $subscriptions-full-width-lg;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      width: $subscriptions-full-width-lg;
    }

    &.payment-method {
      // Use literals here to avoid dark theme from changing the color
      // stylelint-disable-next-line
      background-color: #fff;
      // stylelint-disable-next-line
      color: #303030;
    }
  }

  .gl-form-group,
  .combined {
    flex-grow: 1;
    max-width: 420px;

    legend {
      border-bottom: 0;
      font-weight: $gl-font-weight-bold;
    }

    .gl-form-input,
    .gl-form-select {
      height: 32px;
      line-height: 1rem;
    }
  }

  .number {
    @media(min-width: map-get($grid-breakpoints, lg)) {
      max-width: 202px;
    }
  }

  .label {
    padding: 0;

    .gl-link {
      font-size: inherit;
    }
  }

  .btn {
    line-height: $gl-line-height-14;
  }

  .overview {
    line-height: $gl-line-height-20;

    .btn {
      height: 24px;
      padding: 1px 8px;
      width: 48px;

      @media(min-width: map-get($grid-breakpoints, lg)) {
        width: initial;
      }
    }
  }

  #zuora_payment {
    margin-right: -8px;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      margin-right: 0;
    }

    iframe {
      // Use literals here to avoid dark theme from changing the color
      // stylelint-disable-next-line
      background-color: #fff;
      margin: -4px;
      width: 100% !important;
    }
  }

  .order-summary {
    max-width: $subscriptions-full-width-md;

    @media(min-width: map-get($grid-breakpoints, lg)) {
      max-width: none;
    }
  }

  .edit-group,
  .edit-profile,
  .new-project,
  .verify-identity {
    max-width: 460px;

    .bar {
      width: 100%;
    }

    .normal {
      font-weight: $gl-font-weight-normal;
    }

    .field_with_errors {
      @include gl-flex-grow-1;
      display: inline;
    }

    .project-import {
      h5 {
        display: none;
      }

      .import-buttons {
        justify-content: center;

        .btn {
          width: 288px;
        }
      }
    }
  }
}
