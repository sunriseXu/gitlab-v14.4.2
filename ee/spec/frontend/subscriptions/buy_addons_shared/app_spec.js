import { GlEmptyState, GlAlert } from '@gitlab/ui';
import Vue from 'vue';
import VueApollo from 'vue-apollo';
import { pick } from 'lodash';
import {
  I18N_CI_1000_MINUTES_PLAN,
  I18N_STORAGE_PLAN,
  I18N_API_ERROR,
  planTags,
} from 'ee/subscriptions/buy_addons_shared/constants';
import Checkout from 'ee/subscriptions/buy_addons_shared/components/checkout.vue';
import AddonPurchaseDetails from 'ee/subscriptions/buy_addons_shared/components/checkout/addon_purchase_details.vue';
import OrderSummary from 'ee/subscriptions/buy_addons_shared/components/order_summary.vue';
import SummaryDetails from 'ee/subscriptions/buy_addons_shared/components/order_summary/summary_details.vue';
import App from 'ee/subscriptions/buy_addons_shared/components/app.vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import waitForPromises from 'helpers/wait_for_promises';
import { createMockApolloProvider } from 'ee_jest/subscriptions/spec_helper';
import { mockCiMinutesPlans, mockStoragePlans } from 'ee_jest/subscriptions/mock_data';

Vue.use(VueApollo);

describe('Buy Addons Shared App', () => {
  let wrapper;

  async function createComponent(apolloProvider, injectedProps) {
    wrapper = shallowMountExtended(App, {
      apolloProvider,
      provide: injectedProps,
      stubs: {
        Checkout,
        AddonPurchaseDetails,
        OrderSummary,
        SummaryDetails,
      },
    });
    await waitForPromises();
  }

  const getStoragePlan = () => pick(mockStoragePlans[0], ['id', 'code', 'pricePerYear', 'name']);
  const getCiMinutesPlan = () =>
    pick(mockCiMinutesPlans[0], ['id', 'code', 'pricePerYear', 'name']);
  const findCheckout = () => wrapper.findComponent(Checkout);
  const findOrderSummary = () => wrapper.findComponent(OrderSummary);
  const findAlert = () => wrapper.findComponent(GlAlert);
  const findEmptyState = () => wrapper.findComponent(GlEmptyState);
  const findPriceLabel = () => wrapper.findByTestId('price-per-unit');
  const findQuantityText = () => wrapper.findByTestId('addon-quantity-text');
  const findRootElement = () => wrapper.findByTestId('buy-addons-shared');
  const findSummaryLabel = () => wrapper.findByTestId('summary-label');
  const findSummaryTotal = () => wrapper.findByTestId('summary-total');

  afterEach(() => {
    wrapper.destroy();
  });

  describe('Storage', () => {
    const injectedProps = {
      tags: [planTags.STORAGE_PLAN],
      i18n: I18N_STORAGE_PLAN,
    };
    describe('when data is received', () => {
      beforeEach(async () => {
        const plansQueryMock = jest.fn().mockResolvedValue({ data: { plans: mockStoragePlans } });
        const mockApollo = createMockApolloProvider({ plansQueryMock });
        await createComponent(mockApollo, injectedProps);
      });

      it('should display the root element', () => {
        expect(findRootElement().exists()).toBe(true);
        expect(findEmptyState().exists()).toBe(false);
      });

      it('provides the correct props to checkout', () => {
        expect(findCheckout().props()).toMatchObject({
          plan: { ...getStoragePlan, isAddon: true },
        });
      });

      it('provides the correct props to order summary', () => {
        expect(findOrderSummary().props()).toMatchObject({
          plan: { ...getStoragePlan, isAddon: true },
          title: I18N_STORAGE_PLAN.title,
        });
      });

      describe('and an error occurred', () => {
        beforeEach(() => {
          findOrderSummary().vm.$emit('alertError', I18N_API_ERROR);
        });

        it('shows the alert', () => {
          expect(findAlert().props()).toMatchObject({
            dismissible: false,
            variant: 'danger',
          });
          expect(findAlert().text()).toBe(I18N_API_ERROR);
        });
      });
    });

    describe('when data is not received', () => {
      it('should display the GlEmptyState for empty data', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: null }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });

      it('should display the GlEmptyState for empty plans', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: { plans: null } }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });

      it('should display the GlEmptyState for plans data of wrong type', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: { plans: {} } }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });
    });

    describe('when an error is received', () => {
      it('should display the GlEmptyState', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockRejectedValue(new Error('An error happened!')),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });
    });

    describe('labels', () => {
      const plansQueryMock = jest.fn().mockResolvedValue({ data: { plans: mockStoragePlans } });
      it('shows labels correctly for 1 pack', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText(
          'x 10 GB per pack = 10 GB of storage',
        );
        expect(findSummaryLabel().text()).toBe('1 storage pack');
        expect(findSummaryTotal().text()).toBe('Total storage: 10 GB');
        expect(findPriceLabel().text()).toBe('$60 per 10 GB storage pack per year');
        expect(wrapper.text()).toMatchSnapshot();
      });

      it('shows labels correctly for 2 packs', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock }, { quantity: 2 });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText(
          'x 10 GB per pack = 20 GB of storage',
        );
        expect(findSummaryLabel().text()).toBe('2 storage packs');
        expect(findSummaryTotal().text()).toBe('Total storage: 20 GB');
        expect(findPriceLabel().text()).toBe('$60 per 10 GB storage pack per year');
        expect(wrapper.text()).toMatchSnapshot();
      });

      it('does not show labels if input is invalid', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock }, { quantity: -1 });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText('x 10 GB per pack');
        expect(wrapper.text()).toMatchSnapshot();
      });
    });
  });

  describe('CI Minutes', () => {
    const injectedProps = {
      tags: [planTags.CI_1000_MINUTES_PLAN],
      i18n: I18N_CI_1000_MINUTES_PLAN,
    };

    describe('when data is received', () => {
      beforeEach(async () => {
        const plansQueryMock = jest.fn().mockResolvedValue({ data: { plans: mockCiMinutesPlans } });
        const mockApollo = createMockApolloProvider({ plansQueryMock });
        await createComponent(mockApollo, injectedProps);
      });

      it('should display the root element', () => {
        expect(findRootElement().exists()).toBe(true);
        expect(findEmptyState().exists()).toBe(false);
      });

      it('provides the correct props to checkout', () => {
        expect(findCheckout().props()).toMatchObject({
          plan: { ...getCiMinutesPlan, isAddon: true },
        });
      });

      it('provides the correct props to order summary', () => {
        expect(findOrderSummary().props()).toMatchObject({
          plan: { ...getCiMinutesPlan, isAddon: true },
          title: I18N_CI_1000_MINUTES_PLAN.title,
        });
      });
    });

    describe('when data is not received', () => {
      it('should display the GlEmptyState for empty data', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: null }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });

      it('should display the GlEmptyState for empty plans', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: { plans: null } }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });

      it('should display the GlEmptyState for plans data of wrong type', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockResolvedValue({ data: { plans: {} } }),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });
    });

    describe('when an error is received', () => {
      it('should display the GlEmptyState', async () => {
        const mockApollo = createMockApolloProvider({
          plansQueryMock: jest.fn().mockRejectedValue(new Error('An error happened!')),
        });
        await createComponent(mockApollo, injectedProps);

        expect(findRootElement().exists()).toBe(false);
        expect(findEmptyState().exists()).toBe(true);
      });
    });

    describe('labels', () => {
      const plansQueryMock = jest.fn().mockResolvedValue({ data: { plans: mockCiMinutesPlans } });
      it('shows labels correctly for 1 pack', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText(
          'x 1,000 minutes per pack = 1,000 CI minutes',
        );
        expect(findSummaryLabel().text()).toBe('1 CI minute pack');
        expect(findSummaryTotal().text()).toBe('Total minutes: 1,000');
        expect(findPriceLabel().text()).toBe('$10 per pack of 1,000 minutes');
        expect(wrapper.text()).toMatchSnapshot();
      });

      it('shows labels correctly for 2 packs', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock }, { quantity: 2 });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText(
          'x 1,000 minutes per pack = 2,000 CI minutes',
        );
        expect(findSummaryLabel().text()).toBe('2 CI minute packs');
        expect(findSummaryTotal().text()).toBe('Total minutes: 2,000');
        expect(findPriceLabel().text()).toBe('$10 per pack of 1,000 minutes');
        expect(wrapper.text()).toMatchSnapshot();
      });

      it('does not show labels if input is invalid', async () => {
        const mockApollo = createMockApolloProvider({ plansQueryMock }, { quantity: -1 });
        await createComponent(mockApollo, injectedProps);

        expect(findQuantityText().text()).toMatchInterpolatedText('x 1,000 minutes per pack');
        expect(wrapper.text()).toMatchSnapshot();
      });
    });
  });
});
