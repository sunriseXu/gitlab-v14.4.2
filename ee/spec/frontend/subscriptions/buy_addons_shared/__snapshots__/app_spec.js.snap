// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buy Addons Shared App CI Minutes labels does not show labels if input is invalid 1`] = `
"CI minute packs are only used after you've used your subscription's monthly quota. The additional minutes will roll over month to month and are valid for one year.
     
      CI minute pack
      
            x 1,000 minutes per pack
            
              -1 CI minute packs
             Total minutes: -1,000    Gitlab Org's CI minutes 
        -
       
      CI minutes pack
       
      -
     
        $10 per pack of 1,000 minutes
        Subtotal 
        -
        – Total 
      -
     
      Gitlab Org's CI minutes
     
      CI minutes pack
       
      -
     
        $10 per pack of 1,000 minutes
        Subtotal 
        -
        – Total 
      -"
`;

exports[`Buy Addons Shared App CI Minutes labels shows labels correctly for 1 pack 1`] = `
"CI minute packs are only used after you've used your subscription's monthly quota. The additional minutes will roll over month to month and are valid for one year.
     
      CI minute pack
      
            x 1,000 minutes per pack =
            1,000 CI minutes
              1 CI minute pack
             Total minutes: 1,000    Gitlab Org's CI minutes 
        $10
       
      CI minutes pack
      (x1) 
      $10
     
        $10 per pack of 1,000 minutes
        Subtotal 
        $10
        – Total 
      $10
     
      Gitlab Org's CI minutes
     
      CI minutes pack
      (x1) 
      $10
     
        $10 per pack of 1,000 minutes
        Subtotal 
        $10
        – Total 
      $10"
`;

exports[`Buy Addons Shared App CI Minutes labels shows labels correctly for 2 packs 1`] = `
"CI minute packs are only used after you've used your subscription's monthly quota. The additional minutes will roll over month to month and are valid for one year.
     
      CI minute pack
      
            x 1,000 minutes per pack =
            2,000 CI minutes
              2 CI minute packs
             Total minutes: 2,000    Gitlab Org's CI minutes 
        $20
       
      CI minutes pack
      (x2) 
      $20
     
        $10 per pack of 1,000 minutes
        Subtotal 
        $20
        – Total 
      $20
     
      Gitlab Org's CI minutes
     
      CI minutes pack
      (x2) 
      $20
     
        $10 per pack of 1,000 minutes
        Subtotal 
        $20
        – Total 
      $20"
`;

exports[`Buy Addons Shared App Storage labels does not show labels if input is invalid 1`] = `
"Storage packs
      
            x 10 GB per pack
            
              -1 storage packs
             Total storage: -10 GB    Gitlab Org's storage subscription 
        -
       
      Storage pack
       
      -
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        -
        – Total 
      -
     
      Gitlab Org's storage subscription
     
      Storage pack
       
      -
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        -
        – Total 
      -"
`;

exports[`Buy Addons Shared App Storage labels shows labels correctly for 1 pack 1`] = `
"Storage packs
      
            x 10 GB per pack =
            10 GB of storage
              1 storage pack
             Total storage: 10 GB    Gitlab Org's storage subscription 
        $59.67
       
      Storage pack
      (x1) 
      $59.67
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        $59.67
        – Total 
      $59.67
     
      Gitlab Org's storage subscription
     
      Storage pack
      (x1) 
      $59.67
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        $59.67
        – Total 
      $59.67"
`;

exports[`Buy Addons Shared App Storage labels shows labels correctly for 2 packs 1`] = `
"Storage packs
      
            x 10 GB per pack =
            20 GB of storage
              2 storage packs
             Total storage: 20 GB    Gitlab Org's storage subscription 
        $59.67
       
      Storage pack
      (x2) 
      $59.67
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        $59.67
        – Total 
      $59.67
     
      Gitlab Org's storage subscription
     
      Storage pack
      (x2) 
      $59.67
     
        $60 per 10 GB storage pack per year
       
      Jul 6, 2020 - Dec 15, 2022
       Subtotal 
        $59.67
        – Total 
      $59.67"
`;
